### ISSUE TITLE
Adding PCIE driver to kernel

### ISSUE THREAD

**Issue by foreverska** [OP-TEE/optee_os#3008]
I am attempting to write a driver for a PCIE device against the Qemu-armv8a platform (virt machine).  First pass development efforts phys_to_virt would not resolve 0x3eff0000 (PCIE_PIO).  Looking around I found a function I thought would add PCIE to the memory map, register_phys_mem_pgdir.  So I added 0x3eff0000 of size 0x01010000 to MEM_AREA_IO_SEC.  Phys_to_virt now resolves to 0x107f0cf8 (0xcf8 offset for PIO Config Addr) but this causes a  "Core data-abort" when using the address in io_read32.

Is there something I'm missing?

**Comment by lorc** [comment]
Hi @foreverska,

Could you please post memory map reported by the OP-TEE (it dump it during boot if debug is enabled) and the contents of data abort message?

**Comment by foreverska** [comment]
Sure, I believe in this run I changed the map to NSEC IO just to try it.

>D/TC:0 0 add_va_space:602 type RES_VASPACE size 0x00a00000
>D/TC:0 0 add_va_space:602 type SHM_VASPACE size 0x02000000
>D/TC:0 0 dump_mmap_table:738 type TEE_RAM_RX   va 0x0e100000..0x0e149fff pa 0x0e100000..0x0e149fff size 0x0004a000 (smallpg)
>D/TC:0 0 dump_mmap_table:738 type TEE_RAM_RW   va 0x0e14a000..0x0e2fffff pa 0x0e14a000..0x0e2fffff size 0x001b6000 (smallpg)
>D/TC:0 0 dump_mmap_table:738 type SHM_VASPACE  va 0x0e400000..0x103fffff pa 0x00000000..0x01ffffff size 0x02000000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type IO_NSEC      va 0x10400000..0x115fffff pa 0x3ee00000..0x3fffffff size 0x01200000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type IO_SEC       va 0x11600000..0x117fffff pa 0x09000000..0x091fffff size 0x00200000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type RES_VASPACE  va 0x11800000..0x121fffff pa 0x00000000..0x009fffff size 0x00a00000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type NSEC_SHM     va 0x12200000..0x123fffff pa 0x42000000..0x421fffff size 0x00200000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type TA_RAM       va 0x12500000..0x131fffff pa 0x0e300000..0x0effffff size 0x00d00000 (smallpg)


>E/TC:0 0 Core data-abort at address 0x105f0cf8
>E/TC:0 0  esr 0x96000050  ttbr0 0x0e29f000   ttbr1 0x00000000   cidr 0x0
>E/TC:0 0  cpu #0          cpsr 0x600003c4
>E/TC:0 0  x0  00000000105f0cf8 x1  00000000105f0000
>E/TC:0 0  x2  00000000115fffff x3  000000000e14a678
>E/TC:0 0  x4  0000000001200000 x5  000000000000000c
>E/TC:0 0  x6  000000003eff0000 x7  0000000000000020
>E/TC:0 0  x8  000000000e127c44 x9  000000000e146d1c
>E/TC:0 0  x10 000000000e142f18 x11 0000000006d16908
>E/TC:0 0  x12 0000000039d71903 x13 000000000e2a693c
>E/TC:0 0  x14 0000000011800000 x15 0000000000000000
>E/TC:0 0  x16 0000000000000000 x17 0000000000000000
>E/TC:0 0  x18 0000000000000000 x19 0000000080000000
>E/TC:0 0  x20 0000000000000000 x21 000000000e1480ac
>E/TC:0 0  x22 000000000e14809a x23 0000000000000001
>E/TC:0 0  x24 0000000000000002 x25 000000000000000a
>E/TC:0 0  x26 0000000000000000 x27 0000000000000000
>E/TC:0 0  x28 0000000000000000 x29 000000000e2a6990
>E/TC:0 0  x30 000000000e125af0 elr 000000000e125af0
>E/TC:0 0  sp_el0 000000000e2a6990 

Thanks

**Comment by jenswi-linaro** [comment]
The ESR value seems strange. The ISV bit [24] is 0 meaning that the rest of the lower bits are supposed to be RES0. They aren't and if interpreted says it's an  "0b010000 Synchronous External abort, not on translation table walk". Are you able to access the physical memory using "xp" at the qemu console prompt?

**Comment by foreverska** [comment]
Yes, xp at the physical address returns all zeros without protest.  info pci shows a table of pci devices I expected to see as well.

**Comment by jenswi-linaro** [comment]
It seems that the line:
```
D/TC:0 0 dump_mmap_table:738 type IO_NSEC va 0x10400000..0x115fffff pa 0x3ee00000..0x3fffffff size 0x01200000 (pgdir)
```
Covers what you intended to map as secure IO memory.

**Comment by foreverska** [comment]
Yes, during that run I mapped it to NSEC just to make sure there wasn't an issue with there being two IO_SEC entries.  The issue persists mapped as NSEC or SEC.

Here is the issue this morning as it sits:

>D/TC:0 0 verify_special_mem_areas:501 No NSEC DDR memory area defined
>D/TC:0 0 add_va_space:602 type RES_VASPACE size 0x00a00000
>D/TC:0 0 add_va_space:602 type SHM_VASPACE size 0x02000000
>D/TC:0 0 dump_mmap_table:738 type TEE_RAM_RX   va 0x0e100000..0x0e149fff pa 0x0e100000..0x0e149fff size 0x0004a000 (smallpg)
>D/TC:0 0 dump_mmap_table:738 type TEE_RAM_RW   va 0x0e14a000..0x0e2fffff pa 0x0e14a000..0x0e2fffff size 0x001b6000 (smallpg)
>D/TC:0 0 dump_mmap_table:738 type SHM_VASPACE  va 0x0e400000..0x103fffff pa 0x00000000..0x01ffffff size 0x02000000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type IO_SEC       va 0x10400000..0x105fffff pa 0x09000000..0x091fffff size 0x00200000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type IO_SEC       va 0x10600000..0x117fffff pa 0x3ee00000..0x3fffffff size 0x01200000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type RES_VASPACE  va 0x11800000..0x121fffff pa 0x00000000..0x009fffff size 0x00a00000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type NSEC_SHM     va 0x12200000..0x123fffff pa 0x42000000..0x421fffff size 0x00200000 (pgdir)
>D/TC:0 0 dump_mmap_table:738 type TA_RAM       va 0x12500000..0x131fffff pa 0x0e300000..0x0effffff size 0x00d00000 (smallpg)


>E/TC:0 0 
>E/TC:0 0 Core data-abort at address 0x107f0cf8
>E/TC:0 0  esr 0x96000050  ttbr0 0x0e29f000   ttbr1 0x00000000   cidr 0x0
>E/TC:0 0  cpu #0          cpsr 0x600003c4
>E/TC:0 0  x0  00000000107f0cf8 x1  00000000107f0000
>E/TC:0 0  x2  00000000117fffff x3  000000000e14a6a0
>E/TC:0 0  x4  0000000001200000 x5  000000000000000d
>E/TC:0 0  x6  000000003eff0000 x7  0000000000000020
>E/TC:0 0  x8  000000000e127c48 x9  000000000e146d24
>E/TC:0 0  x10 000000000e142f20 x11 00000000f1906769
>E/TC:0 0  x12 00000000113044e7 x13 000000000e2a693c
>E/TC:0 0  x14 0000000011800000 x15 0000000000000000
>E/TC:0 0  x16 0000000000000000 x17 0000000000000000
>E/TC:0 0  x18 0000000000000000 x19 0000000080000000
>E/TC:0 0  x20 0000000000000000 x21 000000000e1480b4
>E/TC:0 0  x22 000000000e1480a2 x23 0000000000000001
>E/TC:0 0  x24 0000000000000002 x25 000000000000000a
>E/TC:0 0  x26 0000000000000000 x27 0000000000000000
>E/TC:0 0  x28 0000000000000000 x29 000000000e2a6990
>E/TC:0 0  x30 000000000e125af4 elr 000000000e125af4
>E/TC:0 0  sp_el0 000000000e2a6990
>D/TC:0 0 get_fault_type:602 [abort] Unhandled fault!

**Comment by foreverska** [comment]
My knowledge of Qemu internals are limited but if you enable DEBUG_UNASSIGNED it spews
`Unassigned mem write 000000003eff0cf8 = 0x80000000`
Which sounds bad...

**Comment by foreverska** [comment]
So what I did is I migrated my PCIe code to another project just to narrow down if it's Qemu or OPTEE.  Unsurprisingly the processor still faulted out when writing to physical address 0x3eff0cf8.  What did surprise me is it would also fault on address 0x0e000cf8, which is secram, a device I know works.  Similar prognosis for writing to the memory of the secure uart controller.  In all these cases this "unassigned mem write" was printed.  But not RAM, GPIO, MMIO or several other memory areas.  Also "unassigned mem write" is not printed during normal operation of OPTEE which uses secram extensively.

Is there initialization code for some devices (ie secram) in the bootloader somewhere?
